acp/acp.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+/*
+Package acp utilizes the sourcehub acp module to bring the functionality to defradb, this package also helps
+avoid the leakage of direct sourcehub references through out the code base, and eases in swapping
+between local embedded use case and a more global on sourcehub use case.
+*/
+
+package acp
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+var (
+	log = logging.MustNewLogger("acp")
+
+	NoACPModule = immutable.None[ACPModule]()
+)
+
+// ACPModule is the interface to all types of access control modules there might exist.
+//
+// An embedded local implementation might restrict the implementation to not make remote calls.
+//
+// A remote implementation might allow remote calls to access the acp module in the cloud.
+// Note: sourcehub acp module will likely need the remote implementation.
+type ACPModule interface {
+
+	// Start initializes the acp module, with persistence data stored under the given folder.
+	//
+	// If previously started in the same path, then recovers / reloads (unless used in-memory).
+	Start(context.Context, string) error
+
+	// Close closes the resources in use by the acp module.
+	Close() error
+
+	// AddPolicy attempts to add the given policy. Detects the format of the policy automatically
+	// by assuming YAML format if JSON validation fails. Upon success a policyID is returned,
+	// otherwise returns error.
+	//
+	// A policy can not be added without a valid creator.
+	AddPolicy(ctx context.Context, creator string, policy string) (string, error)
+
+	// ValidateResourceExistsOnValidDPI consumes a policyID and resource to validate that the
+	// the following conditions are met:
+	// - policyID is a valid argument.
+	// - resource name is a valid argument.
+	// - policyID corresponds to a valid policy that is registered on acp module.
+	// - resource name exists on that registered policy.
+	// - resource is DPI compliant (more information on DPI below).
+	//
+	// An error is returned upon validation failure (i.e. if above rules not met).
+	//
+	// To be a DPI compliant resource, these rules MUST be met:
```

Comment:
question: Is all of this not validated when the resource is created?  And if not, why are you allowed to create invalid resources?

Reply:
What is valid is subjective to the sourcehub (dw still talking about local acp module) acp module.

- Sourcehub has it's own reasons for allow flexible policy types.
- DefraDB has to allow uploading of all policy types that sourcehub deems valid.
- Fun fact a policy might exist without any resources, for whatever purpose sourcehub needs it is irrelevant to DefraDB, Hence DPI compliance, which is a subset of rules that DefraDB cares about.

So to answer you questions:

> Is all of this not validated when the resource is created

If by resource is created you mean, when the policy that has the resource is created, then no, only sourcehub's policy validation happens on it, not DefraDBs (to allow creating other types of policies, this will improve in future with meta policies)

> And if not, why are you allowed to create invalid resources?

If first paragraph didn't make sense then more here:

For other sourcehub functionalities and feature we can't limit it to for example have `owner` on all relations (in fact there is also no notion of `required` relation to exist on required permissions sourcehub policy).

Reply:
Okay I think I understand better, I have a follow up suggestions:

suggestion: Why does this function signature care about the distinction between a non-defra resource existing in sourcehub, and the resource not existing at all?  Could this function not be called `ValidateResourceExists`?

suggestion: ValitateFoExists type functions are often easier to use if they return booleans.  If I want to check whether it exists or not a bool is much nicer than having to handle various errors that might be returned.

question: Does the Defra/`ACPModule` allow the uploading of (defra) invalid policies/resources?  And if so, why?

Reply:
> Okay I think I understand better, I have a follow up suggestions:
> 
> suggestion: Why does this function signature care about the distinction between a non-defra resource existing in sourcehub, and the resource not existing at all? Could this function not be called `ValidateResourceExists`?

It doesn't care about non-defra resources existing on sourcehub, it cares however that the specified resource name is a valid DPI-resource on the specified policy.

No, this name `ValidateResourceExists` is inaccurate because the resource can exist on a policy that has policyID but the resource might not be a valid DPI resource (Note: Just because a resource exists doesn't mean it is a valid DPI resource.)
 
> suggestion: ValitateFoExists type functions are often easier to use if they return booleans. If I want to check whether it exists or not a bool is much nicer than having to handle various errors that might be returned.

- [ ] Might do.


> question: Does the Defra/`ACPModule.AddPolicy` allow the uploading of (defra) invalid policies/resources? And if so, why?

- For other sourcehub functions, and other DefraDB access control type policy functions we might want to have different types of policy standards (will be easier to do this with MetaPolicies when they are implemented).

Reply:
> No, this name ValidateResourceExists is inaccurate because the resource can exist on a policy that has policyID but the resource might not be a valid DPI resource (Note: Just because a resource exists doesn't mean it is a valid DPI resource.)

I understand that, but anyone calling this *defra* function is unlikely to care about the distinction.

> For other sourcehub functions, and other DefraDB access control type policy functions we might want to have different types of policy standards (will be easier to do this with MetaPolicies when they are implemented).

Yes, but why does this ACPModule allow the uploading of policies that are invalid to itself?

Reply:
> I understand that, but anyone calling this _defra_ function is unlikely to care about the distinction.

I disagree, `ValidateResourceExists` is an invalid and incorrect name. Tomorrow this module will add other types of admin operations which might require different types of resources to be valid say "XPI" compliant resources they will all live on this module as they will also be a `_defra_` function. 
 
> > For other sourcehub functions, and other DefraDB access control type policy functions we might want to have different types of policy standards (will be easier to do this with MetaPolicies when they are implemented).
> 
> Yes, but why does this ACPModule allow the uploading of policies that are invalid to itself?

There is only one ACP Module I know of on DefraDB, which is this one, so it allows it. It is only invalid to a sub-portion of access type to itself. Maybe @jsimnz can elaborate more on other use cases why we don't want to limit sourcehub to only accept DPI valid policies.

Reply:
> why we don't want to limit sourcehub to only accept DPI valid policies

I am not suggesting this.  Here we have a defra specific interface, for uploading policies via defra, that only validates valid defra policies/resources.  It feels very strange to me to allow the uploading of invalid defra policies through it, and I suspect that it would cause far more problems for our users than it would ever solve.

Reply:
>  Tomorrow this module will add other types of admin operations which might require different types of resources to be valid say "XPI" compliant resources they will all live on this module as they will also be a _defra_ function.

Do you think it unreasonable to assume that other types of **defra** admin operations would need to be declared using `Defra Policy Interface` documents?

Reply:
> I am not suggesting this. Here we have a defra specific interface, for uploading policies via defra, that only validates valid defra policies/resources. It feels very strange to me to allow the uploading of invalid defra policies through it, and I suspect that it would cause far more problems for our users than it would ever solve.

Went over on ACP Demo call, let me know if any other out standing discussion item on this topic is left.

Reply:
> Do you think it unreasonable to assume that other types of **defra** admin operations would need to be declared using `Defra Policy Interface` documents?

Cant say for sure as it has not been `spec`ed out yet. However if I had to guess, I would say it is more likely than not that the existing DPI would be invalid for the admin operations (i could be wrong though).

Reply:
> Went over on ACP Demo call, let me know if any other out standing discussion item on this topic is left.

Yes - in the demo it was mentioned:

1. A DPI is not valid until it is linked to a defra collection, and this was why validation could not take place on policy upload (as the collection might not exist yet) - but I do not see any mention of this in the documentation.  Can you please clarify/expand the documentation?
2. In the demo we mentioned the need for users to be able to get uploaded policies regardless of whether they are valid DPIs or not - I see this as very important to cover the use case of 'upload policy', 'my bad, I'm not sure what I uploaded, lets check'.

Reply:
> Yes - in the demo it was mentioned:
> 
> 1. A DPI is not valid until it is linked to a defra collection, and this was why validation could not take place on policy upload (as the collection might not exist yet) - but I do not see any mention of this in the documentation.  Can you please clarify/expand the documentation?


I am not sure what you mean by "A DPI is not valid until it is linked to a defra collection". DPI means to be compliant so a DPI by definition can not be invalid. What was mentioned in the demo is that when a policy is linked during the `client schema add` command, with a "policy id" and "resource name" specified on the directive, then only the resource name specified within the target policy (policy that corresponds to that specified policy id within the acp module) is validated to ensure it follows the DPI compliance.

This is not true "and this was why validation could not take place on policy upload". We can validate the policy at upload step by validating ALL resources in a brute force way to ensure the entire policy is DPI compliant, but we choose not to because of the following 3 reasons:
1) In reality sourcehub has no requirement of DPI compliance and can contain many types of policies.
2) Due to (1) we have to validate at the linking step anyways.
3) At the policy upload step we have no information about which "resource" actually required the DPI compliance as it can have many resources that aren't DPI compliant.

Side thought: We could in future introduce a `--strict` flag if the users want to validate all the resources at upload step for their own sake, even if it makes no sense from the perspective of sourcehub, but validation at schema add step is still a must.

Reply:
> 2. In the demo we mentioned the need for users to be able to get uploaded policies regardless of whether they are valid DPIs or not - I see this as very important to cover the use case of 'upload policy', 'my bad, I'm not sure what I uploaded, lets check'.

Discussed this with @Lodek and we can dump the uploaded policy once sourcehub introduces the capability through a client function. I made an issue to track this: https://github.com/sourcenetwork/defradb/issues/2388 

However, I see that as out of scope for this PR.

Reply:
> We can validate the policy at upload step by validating ALL resources in a brute force way to ensure the entire policy is DPI compliant

In the demo it was suggested that this was not possible, and that was the reason why I considered this discussion closed.  If it is possible then I think we should do it.

This is a Defra interface, and if we can validate on upload we should. It's much much easier for users if they know they have messed up on upload than after.

Reply:
> In the demo it was suggested that this was not possible, and that was the reason why I considered this discussion closed. If it is possible then I think we should do it.
> 
> This is a Defra interface, and if we can validate on upload we should. It's much much easier for users if they know they have messed up on upload than after.

Yes, we don't know the exact resource they will use at that time so we can't validate but what I am mentioning is that a dumb option might be to validate all resources in a brute force way if a user might want that. However, I now lean towards not allowing that as I thought about what if a user wants to use a resource that is not DPI compliant under the permissions expression to inherit its permissions (this will soon be implemented and supported automagically as source supports it) so, in that case, the extra validation would cause a hinderance. Please ignore my brute-force suggestion of validation above.

Reply:
I'm not fully following what you mean here, but it sounds like you are still thinking things through :)

RE 'brute force', why the reluctance here?  To me this is not a performance sensitive space and I'd probably chose even an extra couple of seconds of computation over weaker validation.

Reply:
> I'm not fully following what you mean here, but it sounds like you are still thinking things through :)

I am haha

> RE 'brute force', why the reluctance here? To me this is not a performance sensitive space and I'd probably chose even an extra couple of seconds of computation over weaker validation.

It's not about the performance at all, I wouldn't worry about performance here. The problem at hand is that it would break features for users in the near future as they actually might have resources that are not-DPI compliant that won't be used on the collection with a `@policy` command but might serve another valid purpose, like being inherited into other policies that are DPI compliant. Might get a more clear idea in the near future on things we can improve on, but for now on yea it's impossible to know the policy resource the user will target and hence we will do validation on the linking step only.

Reply:
> like being inherited into other policies that are DPI compliant.

Ah I think it is becoming clearer to me - it might be worth documenting some of this on `AddPolicy` -  at the moment there is no documentation on what gets validated at the time of uploading, and why which bits are and aren't validated.

Reply:
- [ ] document better